node/network/availability-distribution/src/lib.rs

```diff
@@ -0,0 +1,1093 @@
+			RuntimeApiRequest::CandidatePendingAvailability(para, tx),
+		)))
+		.await?;
+	rx.await?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Query the validator set.
+async fn query_validators<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+) -> Result<Vec<ValidatorId>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	let query_validators = AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+		relay_parent,
+		RuntimeApiRequest::Validators(tx),
+	));
+
+	ctx.send_message(query_validators)
+		.await?;
+	rx.await?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Query the hash of the `K` ancestors
+async fn query_k_ancestors<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+	k: usize,
+) -> Result<Vec<Hash>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	let query_ancestors = AllMessages::ChainApi(ChainApiMessage::Ancestors {
+		hash: relay_parent,
+		k,
+		response_channel: tx,
+	});
+
+	ctx.send_message(query_ancestors)
+		.await?;
+	rx.await?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Query the session index of a relay parent
+async fn query_session_index_for_child<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+) -> Result<SessionIndex>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	let (tx, rx) = oneshot::channel();
+	let query_session_idx_for_child = AllMessages::RuntimeApi(RuntimeApiMessage::Request(
+		relay_parent,
+		RuntimeApiRequest::SessionIndexForChild(tx),
+	));
+
+	ctx.send_message(query_session_idx_for_child)
+		.await?;
+	rx.await?
+		.map_err::<Error, _>(Into::into)
+}
+
+/// Queries up to k ancestors with the constraints of equiv session
+async fn query_up_to_k_ancestors_in_same_session<Context>(
+	ctx: &mut Context,
+	relay_parent: Hash,
+	k: usize,
+) -> Result<Vec<Hash>>
+where
+	Context: SubsystemContext<Message = AvailabilityDistributionMessage>,
+{
+	// k + 1 since we always query the child's session index
+	// ordering is [parent, grandparent, greatgrandparent, greatgreatgrandparent, ...]
+	let ancestors = query_k_ancestors(ctx, relay_parent, k + 1).await?;
+	let desired_session = query_session_index_for_child(ctx, relay_parent).await?;
+	// we would only need `ancestors.len() - 1`, but the one extra could avoid a re-alloc
+	// if the consumer wants to push the `relay_parent` onto it too and does not hurt otherwise
+	let mut acc = Vec::with_capacity(ancestors.len());
+
+	// iterate from youngest to oldest
+	let mut iter = ancestors.into_iter().peekable();
+
+	while let Some(ancestor) = iter.next() {
+		if let Some(ancestor_parent) = iter.peek() {
+			let session = query_session_index_for_child(ctx, *ancestor_parent).await?;
+			if session != desired_session {
+				break;
+			}
+			if ancestor != relay_parent {
+				acc.push(ancestor);
+			}
+		} else {
+			// we reached the genesis, which does not have ancestors
```

Comment:
not necessarily the genesis, but the oldest block in `ancestors`

Reply:
In this case it must be genesis, since we are only interested in `k` blocks session index, but we query `k+1`.
Assuming the case of `k` being chosen with a good error margin around the start-to-prune history depth, this should only happen in the genesis case.

Reply:
Actually we don't want to include the genesis block here, there is nothing to do for it anyways.

Reply:
Good point, I don't know if it's guaranteed that if we good pruning depth, query will never return less than `k+1` and `>0` hashes. Could it be the case that tip of the chain and relay parent we're querying are far away from each other. This is all speculation atm, but it would be good to handle this case in case our assumption is not correct.